csharp/ql/src/semmle/code/csharp/dataflow/LibraryTypeDataFlow.qll

```diff
@@ -1679,6 +1679,62 @@ class SystemCollectionsGenericKeyValuePairStructFlow extends LibraryTypeDataFlow
   }
 }
 
+/** Data flow for `System.[Value]Tuple<,...,>`. */
+class SystemTupleFlow extends LibraryTypeDataFlow, ValueOrRefType {
+  SystemTupleFlow() {
+    this.getNamespace() instanceof SystemNamespace and
+    this.getName().regexpMatch("(Value)?Tuple(<,*>)?")
+    or
+    this instanceof TupleType
+  }
+
+  private AccessPath getItemAccessPath(int i) {
+    i in [1 .. count(this.getAMember())] and
+    result in [AccessPath::field(this.getField("Item" + i)),
+          AccessPath::property(this.getProperty("Item" + i))]
+  }
+
+  override predicate callableFlow(
+    CallableFlowSource source, AccessPath sourceAp, CallableFlowSink sink, AccessPath sinkAp,
+    SourceDeclarationCallable c, boolean preservesValue
+  ) {
+    preservesValue = true and
+    (
+      exists(SystemTupleFlow t, int i |
+        source = getFlowSourceArg(c, i - 1, _) and
+        sourceAp = AccessPath::empty() and
+        sink = TCallableFlowSinkReturn() and
+        sinkAp = t.getItemAccessPath(i)
+      |
+        c.(Constructor).getDeclaringType() = this and
+        t = this
+        or
+        c = this.getAMethod(any(string name | name.regexpMatch("Create(<,*>)?"))) and
+        t = c.getReturnType().getSourceDeclaration()
+      )
+      or
+      c =
+        any(ExtensionMethod m |
```

Comment:
This is a misuse of the language, but you can do the following:
```
using System;
public static class C {
    
    public static void Deconstruct<T1, T2, T3>(
        this ValueTuple<T1, T2, T3> p, 
        out T1 f, 
        out int whatever)
    {
        f = p.Item1;
        whatever = 42;
    }
    
    public static void Main() 
    {
        C.Deconstruct<string, string, string>(
            ("a", "b", "c"), 
            out var f, 
            out var w);
        Console.WriteLine(f);   
        Console.WriteLine(w);  
    }
}
```
Here, `C.Deconstruct` is a `Deconstruct ` extension method on `ValueTuple`, but the `out` parameters are disconnected from the `Item` properties.

Reply:
This also works with deconstruction syntax:
```
using System;
public static class C {
    
    public static void Deconstruct<T1, T2, T3>(
        this Tuple<T1, T2, T3> p, 
        out T1 f, 
        out int whatever)
    {
        f = p.Item1;
        whatever = 42;
    }
    
    public static void Main() 
    {
        var t = Tuple.Create("a", "b", "c");
        var (f, w) = t;
        Console.WriteLine(f);   
        Console.WriteLine(w);  
    }
}
```

Reply:
Yes, but I think it is OK to assume that any `Deconstruct` method will do "the right thing".